tsdb/tsdb.go

```diff
@@ -506,90 +505,79 @@ func (c *Chain) GetPositionalPointerNormalized(n int64) (*Block, error) {
 	c.mux.Lock()
 	defer c.mux.Unlock()
 	c.LengthElements = len(c.Chain)
-	jumpSize := int(math.Floor(math.Sqrt(float64(c.LengthElements))))
-	index := jumpSize - 1
-
 	if c.Chain[c.LengthElements-1].NormalizedTime < n || c.Chain[0].NormalizedTime > n {
 		return nil, errors.New("Normalized time not in Chain range")
 	}
+	l, r, m := 0, c.LengthElements, 0
 
-	for n > c.Chain[index].NormalizedTime && index < c.LengthElements-jumpSize {
-		index += jumpSize
-	}
-
-	for c.Chain[index].NormalizedTime < n {
-		index++
-	}
+	for l <= r {
+		m = l + (r-l)/2
+		if c.Chain[m].NormalizedTime == n {
+			return &c.Chain[m], nil
+		}
 
-	for c.Chain[index].NormalizedTime > n {
-		index--
-	}
+		if c.Chain[m].NormalizedTime < n {
+			l = m + 1
+		} else {
+			r = m - 1
+		}
 
-	if c.Chain[index].NormalizedTime != n {
-		return nil, errors.New("Normalized time not found in chain")
 	}
 
-	return &c.Chain[index], nil
+	return nil, errors.New("Normalized time not found in chain")
 }
 
 // GetPositionalPointerNormalizedPing Returns block by searching the chain for the NormalizedTime
 func (c *ChainPing) GetPositionalPointerNormalizedPing(n int64) (*BlockPing, error) {
 	c.mux.Lock()
 	defer c.mux.Unlock()
 	c.LengthElements = len(c.Chain)
-	jumpSize := int(math.Floor(math.Sqrt(float64(c.LengthElements))))
-	index := jumpSize - 1
 
 	if c.Chain[c.LengthElements-1].NormalizedTime < n || c.Chain[0].NormalizedTime > n {
 		return nil, errors.New("Normalized time not in Chain range")
 	}
+	l, r, m := 0, c.LengthElements, 0
 
-	for n > c.Chain[index].NormalizedTime && index < c.LengthElements-jumpSize {
-		index += jumpSize
-	}
-
-	for c.Chain[index].NormalizedTime < n {
-		index++
-	}
+	for l <= r {
+		m = l + (r-l)/2
+		if c.Chain[m].NormalizedTime == n {
+			return &c.Chain[m], nil
+		}
```

Comment:
I think the infinite loop is not being stopped in cases of an invalid query, or a block not found. This will lead to infinite loop, thus freezing the app. I am correct?

Reply:
the loop will terminate due to a condition in the for loop  ,that is ,l<=r